### ISSUE TITLE
Option to treat a success (zero) exit code as failure

### ISSUE THREAD

**Issue by dechamps** [systemd/systemd#10297]
**Is your feature request related to a problem? Please describe.**

Many services (daemons) are never supposed to exit unless explicitly stopped using `systemctl stop`, period. If the service process exits with a success or success-like exit code (e.g. zero, `SIGINT`, etc.), it's unlikely that this is actually a normal condition; it's way more likely the process erroneously returned a success code instead of a failure code.

Nevertheless, systemd will still treat a process exiting with a success code as a normal exit. It is not possible to treat success codes as errors. Therefore, any monitoring system that watches for unit states to surface problems (e.g. using `systemctl is-system-running`) will be unable to catch this failure mode.

This recently happened to me as my [deluge](https://deluge-torrent.org/) daemons ran into [this issue](https://forum.deluge-torrent.org/viewtopic.php?t=55059): it turns out that this particular error made deluge exit with a success code, not an error code. It took me days to notice that my deluge daemons broke, because as far as systemd was concerned, the daemons exited normally and everything was fine.

**Describe the solution you'd like**

The [`SuccessExitStatus`](https://www.freedesktop.org/software/systemd/man/systemd.service.html#SuccessExitStatus=) option comes very close to solving this problem, but sadly it looks like it only allows adding new success codes, not removing from the default list (which includes exit code 0). If there was a way to write something like, say, `SuccessExitStatus= -0`, it would solve the problem.

**Describe alternatives you've considered**

I'm currently using a workaround where I wrap the daemon process inside a script that alters the return code. This works, but it feels hacky, and looking at existing systemd features like `SuccessExitStatus` it clearly looks like something that systemd should be able to do natively.

**Comment by coretemp** [comment]
Broken daemons should be fixed. They should not give rise to opening issues against an unrelated piece of software (systemd).

It appears to be really popular to just expect broken programs to allow to exist and have systemd provide workarounds for everything.

**Comment by dechamps** [comment]
> Broken daemons should be fixed.

I agree. This is not mutually exclusive with this feature request, though. Requiring that all software running under systemd be bug-free is not realistic. systemd has never made this assumption, and presumably never will.

> It appears to be really popular to just expect broken programs to allow to exist and have systemd systemd provide workarounds for everything.

I believe it is a good thing that systemd provides "defense in depth" mechanisms that increase resiliency to bugs and failures. For example, systemd already provides watchdog functionality for catching issues with services getting stuck, and options for automatically restarting services that crash. I am merely requesting that systemd catches an additional failure mode (services accidentally exiting with the wrong code) that is quite trivial in comparison.

I do agree that such behaviour should not be enabled by default, because that would indeed create the wrong incentives and allow bugs to go unnoticed. However, an explicit option seems like a good tradeoff to me.

**Comment by coretemp** [comment]
First of all I don't think all decisions made in systemd are good decisions and I don't consider the main author to be a light of perfect judgment (just look at the Wikipedia page for systemd for many others to agree with this). It's just that the average person requesting features here has absolutely zero clue compared to him. I suppose it's very nice of him that he even chooses to talk to so many people with horrible ideas (I just looked for a couple of hours at these issues and I am already tired of the displayed incompetence and ignorance). 

Second of all, I completely disagree with your "defense in depth" idea. All it does is provide excuses for people who cannot write software to claim that it does work or to delay work on those services. If anything, it would probably be better if servers would catch fire when there was even a single bug in them.

At least then it would be clear for management which systems are actually working and which ones are not. It would probably even be cheaper in the long term (you need to have worked in a special environment before you can appreciate this).

**Comment by keszybz** [comment]
I guess the syntax should be something like `SuccessExitStatus=~0 SIGTERM`, to explicitly remove 0 and SIGTERM from the list of successful exit codes.

**Comment by dustymabe** [comment]
I ran across this today. My desired case is a `.service` unit where if the command I'm running succeeds (`0` exit code) I want the unit to fail.

I can get this with something like `bash -c "true && exit 1 || exit 0"`, but it's not ideal.

**Comment by dechamps** [comment]
I believe I managed to come up with a slightly cleaner workaround that does not require wrapping the service process. Simply add:

```
ExecStopPost=/usr/bin/sh -c 'test "$EXIT_CODE" != exited'
```

Basically, if the service stops running for any reason, this runs a command that will deliberately fail if `$EXIT_CODE` is `exited`. This will fail the service if the main service process just exits normally. It will not fail the service if it is stopped through `systemctl stop`, because in that case `$EXIT_CODE` is `killed`.
